channeldb/messagestore.go

```diff
@@ -0,0 +1,164 @@
+package channeldb
+
+import (
+	"bytes"
+	"errors"
+
+	"github.com/coreos/bbolt"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// messageStoreBucket is a key used to create a top level bucket in the
+	// gossiper database, used for storing messages that are to be sent to
+	// peers. Upon restarts, these messages will be read and resent to their
+	// respective peers.
+	//
+	// maps:
+	//   pubKey (33 bytes) + msgShortChanID (8 bytes) + msgType (2 bytes) -> msg
+	messageStoreBucket = []byte("message-store")
+
+	// ErrUnsupportedMessage is an error returned when we attempt to add a
+	// message to the store that is not supported.
+	ErrUnsupportedMessage = errors.New("unsupported message type")
+)
+
+// GossipMessageStore is a store responsible for storing gossip messages which
+// we should reliably send to our peers. By design, this store will only keep
+// the latest version of a message (like in the case of multiple
+// ChannelUpdate's) for a channel with a peer.
+type GossipMessageStore struct {
+	db *DB
+}
+
+// NewMessageStore creates a new message store backed by a channeldb instance.
+func NewMessageStore(db *DB) *GossipMessageStore {
+	return &GossipMessageStore{db}
+}
+
+// messageStoreKey constructs the database key for the message to be stored.
+func messageStoreKey(msg lnwire.Message, peerPubKey [33]byte) ([]byte, error) {
+	var shortChanID lnwire.ShortChannelID
+	switch msg := msg.(type) {
+	case *lnwire.AnnounceSignatures:
+		shortChanID = msg.ShortChannelID
+	case *lnwire.ChannelUpdate:
+		shortChanID = msg.ShortChannelID
+	default:
+		return nil, ErrUnsupportedMessage
+	}
+
+	var k [33 + 8 + 2]byte
+	copy(k[:33], peerPubKey[:])
+	byteOrder.PutUint64(k[33:41], shortChanID.ToUint64())
+	byteOrder.PutUint16(k[41:43], uint16(msg.MsgType()))
+
+	return k[:], nil
+}
+
+// AddMessage adds a message to the store for this peer.
+func (s *GossipMessageStore) AddMessage(msg lnwire.Message,
+	peerPubKey [33]byte) error {
+
+	// Construct the key for which we'll find this message with in the store.
+	msgKey, err := messageStoreKey(msg, peerPubKey)
+	if err != nil {
+		return err
+	}
+
+	// Serialize the message with its wire encoding.
+	var b bytes.Buffer
+	if _, err := lnwire.WriteMessage(&b, msg, 0); err != nil {
+		return err
+	}
+
+	return s.db.Update(func(tx *bbolt.Tx) error {
+		messageStore, err := tx.CreateBucketIfNotExists(messageStoreBucket)
```

Comment:
would prefer this to be created on initialization of the store, then from then on we can assume it exists

Reply:
Fixed.